File: src/main/java/com/dec/decisland/DecIslandModDataGenerator.java
package com.dec.decisland;

import com.dec.decisland.datagen.*;
import net.minecraft.data.loot.LootTableProvider;
import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.data.event.GatherDataEvent;

import java.util.List;
import java.util.Set;

@EventBusSubscriber(modid = DecIsland.MOD_ID)
public class DecIslandModDataGenerator {
    @SubscribeEvent
    public static void gatherData(GatherDataEvent.Client event) {
        event.createProvider(ModModelsProvider::new);
        event.createProvider(ModDataMapProvider::new);

        String[] locales = {"en_us", "zh_cn"};
        for (String locale : locales) {
            event.createProvider(output -> new ModLangProvider(output, locale) {
            });
        }

        event.createProvider(((output, lookupProvider) ->
                new LootTableProvider(output, Set.of(), List.of(new LootTableProvider.SubProviderEntry(
                        ModBlockLootTablesProvider::new, LootContextParamSets.BLOCK
                )), lookupProvider)));

        event.createProvider((ModBlockTagsProvider::new));
        event.createProvider((ModItemTagsProvider::new));
    }
}

File: src/main/java/com/dec/decisland/DecIslandModDataGenerator.kt
package com.dec.decisland

import com.dec.decisland.datagen.ModBlockLootTablesProvider
import com.dec.decisland.datagen.ModBlockTagsProvider
import com.dec.decisland.datagen.ModDataMapProvider
import com.dec.decisland.datagen.ModItemTagsProvider
import com.dec.decisland.datagen.ModLangProvider
import com.dec.decisland.datagen.ModModelsProvider
import net.minecraft.data.loot.LootTableProvider
import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.neoforge.data.event.GatherDataEvent

@EventBusSubscriber(modid = DecIsland.MOD_ID)
object DecIslandModDataGenerator {
    @SubscribeEvent
    @JvmStatic
    fun gatherData(event: GatherDataEvent.Client) {
        event.createProvider(::ModModelsProvider)
        event.createProvider(::ModDataMapProvider)

        arrayOf("en_us", "zh_cn").forEach { locale ->
            event.createProvider { output ->
                object : ModLangProvider(output, locale) {}
            }
        }

        event.createProvider { output, lookupProvider ->
            LootTableProvider(
                output,
                setOf(),
                listOf(LootTableProvider.SubProviderEntry(::ModBlockLootTablesProvider, LootContextParamSets.BLOCK)),
                lookupProvider,
            )
        }

        event.createProvider(::ModBlockTagsProvider)
        event.createProvider(::ModItemTagsProvider)
    }
}
